### PR TITLE
Variable Cost Refactor Part 1: Function Data

### PR DESCRIPTION
**Pull request by GabrielKS**
The first part of a significant refactor of Sienna cost data structures. Relates to https://github.com/NREL-Sienna/PowerSystems.jl/issues/934, https://github.com/NREL-Sienna/PowerSystems.jl/issues/985, https://github.com/NREL-Sienna/PowerSystems.jl/issues/1007.

At the time of draft PR creation, all tests pass except for these:
  * `Test field validation` in `test_validation.jl`: an issue unrelated to these changes, a fix will become available on `psy4` soon and I'll rebase onto that
  * Several tests in `test_cost_functions.jl`: waiting on a decision about whether to refactor aspects of InfrastructureSystems (e.g., the `Deterministic` struct)

…provided that [this commit](https://github.com/NREL-Sienna/PowerSystemsTestData/tree/7910da87f7749f1baa17286315542bc593e16605) of PowerSystemsTestData is used (by changing the PowerSystemsCaseBuilder `DATA_DIR` to a local clone of that).

This is not quite ready for full review but I'm putting it out here to track progress.

### CODE REVIEW

File: src/PowerSystems.jl [hunk 1]
```diff
@@ -39,14 +39,25 @@ export Transformer2W
 export TapTransformer
 export PhaseShiftingTransformer
 
+export FunctionData
+export LinearFunctionData
+export QuadraticFunctionData
+export PolynomialFunctionData
+export PiecewiseLinearPointData
+export PiecewiseLinearSlopeData
+export get_proportional_term, get_quadratic_term, get_constant_term, get_coefficients
```
Comment (jd-lara) [inline src/PowerSystems.jl:48]:
```suggestion
export get_proportional_term 
export get_quadratic_term 
export get_constant_term 
export get_coefficients
```

File: src/parsers/power_system_table_data.jl [hunk 2]
```diff
@@ -855,10 +855,11 @@ function calculate_variable_cost(
         for i in 2:length(var_cost)
             var_cost[i] = (var_cost[i - 1][1] + var_cost[i][1], var_cost[i][2])
         end
+    var_cost = PiecewiseLinearPointData([(x, y) for (y, x) in var_cost])
```
[the remark this answers is not in the document]
Comment (jd-lara) [reply src/parsers/power_system_table_data.jl:858]: Formatter run missing

File: test/test_cost_functions.jl [hunk 3]
```diff
@@ -1,94 +1,88 @@
...
-    generator = get_component(ThermalStandard, sys, get_name(generators[1]))
-    market_bid = MarketBidCost(nothing)
-    set_operation_cost!(generator, market_bid)
-    forecast = IS.Deterministic("variable_cost", data_polynomial, resolution)
-    set_variable_cost!(sys, generator, forecast)
-    for s in generator.services
-        forecast = IS.Deterministic(get_name(s), service_data, resolution)
-        set_service_bid!(sys, generator, s, forecast)
-    end
+# TODO figure out how to handle costs in InfrastructureSystems
```
Comment (jd-lara) [inline test/test_cost_functions.jl:8]: This is more about the implementation of MarketBidCost that anything else.

File: src/PowerSystems.jl (hunk 1 above)
Comment (GabrielKS) [reply src/PowerSystems.jl]: Done

**Review by jd-lara — CHANGES_REQUESTED** [review]
After an offline conversation, the decision is to move the code for FunctionData to IS and re-export in PSY. 

Implement that change in this PR prior to merging.

**Comment by GabrielKS** [comment]
Rebased on top of https://github.com/NREL-Sienna/PowerSystems.jl/pull/1057

**Comment by GabrielKS** [comment]
Never mind about https://github.com/NREL-Sienna/PowerSystems.jl/pull/1057, it was a bad idea. All those changes are here.

**Comment by GabrielKS** [comment]
Finally ready for review! Now that the FunctionData structs live in InfrastructureSystems, this PR is mostly adapting existing cost data structures to use the new ones, plus System JSON conversion code.

IS, PSY, and PSB tests pass when using the appropriate branches of each plus PowerSystemsTestData:
Screenshot of passing tests

**Comment by GabrielKS** [comment]
The group of four PRs this is a part of:

1. https://github.com/NREL-Sienna/InfrastructureSystems.jl/pull/331
2. https://github.com/NREL-Sienna/PowerSystems.jl/pull/1056
3. https://github.com/NREL-Sienna/PowerSystemCaseBuilder.jl/pull/72
4. https://github.com/NREL-Sienna/PowerSystemsTestData/pull/50

**Review by jd-lara — APPROVED** [review]

**Pull request merged**